shell/platform/embedder/embedder_surface_vulkan.cc

```diff
@@ -0,0 +1,148 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/embedder/embedder_surface_vulkan.h"
+
+#include "flutter/shell/common/shell_io_manager.h"
+#include "include/gpu/GrDirectContext.h"
+#include "include/gpu/vk/GrVkBackendContext.h"
+#include "include/gpu/vk/GrVkExtensions.h"
+#include "shell/gpu/gpu_surface_vulkan.h"
+#include "shell/gpu/gpu_surface_vulkan_delegate.h"
+
+namespace flutter {
+
+EmbedderSurfaceVulkan::EmbedderSurfaceVulkan(
+    uint32_t version,
+    VkInstance instance,
+    size_t instance_extension_count,
+    const char** instance_extensions,
+    size_t device_extension_count,
+    const char** device_extensions,
+    VkPhysicalDevice physical_device,
+    VkDevice device,
+    uint32_t queue_family_index,
+    VkQueue queue,
+    VulkanDispatchTable vulkan_dispatch_table,
+    std::shared_ptr<EmbedderExternalViewEmbedder> external_view_embedder)
+    : vk_(fml::MakeRefCounted<vulkan::VulkanProcTable>(
+          vulkan_dispatch_table.get_instance_proc_address)),
+      device_(*vk_,
+              vulkan::VulkanHandle<VkPhysicalDevice>{physical_device},
+              vulkan::VulkanHandle<VkDevice>{device},
+              queue_family_index,
+              vulkan::VulkanHandle<VkQueue>{queue}),
+      vulkan_dispatch_table_(vulkan_dispatch_table),
+      external_view_embedder_(external_view_embedder) {
+  // Make sure all required members of the dispatch table are checked.
+  if (!vulkan_dispatch_table_.get_instance_proc_address ||
+      !vulkan_dispatch_table_.get_next_image ||
+      !vulkan_dispatch_table_.present_image) {
+    return;
+  }
+
+  vk_->SetupInstanceProcAddresses(vulkan::VulkanHandle<VkInstance>{instance});
+  vk_->SetupDeviceProcAddresses(vulkan::VulkanHandle<VkDevice>{device});
+  if (!vk_->IsValid()) {
+    FML_LOG(ERROR) << "VulkanProcTable invalid.";
+    return;
+  }
+
+  main_context_ = CreateGrContext(instance, version, instance_extension_count,
+                                  instance_extensions, device_extension_count,
+                                  device_extensions, ContextType::kRender);
+  // TODO(bdero): Add a second (optional) queue+family index to the Embedder API
+  //             to allow embedders to specify a dedicated transfer queue for
+  //             use by the resource context. Queue families with graphics
+  //             capability can always be used for memory transferring, but it
+  //             would be adventageous to use a dedicated transter queue here.
+  resource_context_ = CreateGrContext(
+      instance, version, instance_extension_count, instance_extensions,
+      device_extension_count, device_extensions, ContextType::kResource);
+
+  valid_ = main_context_ && resource_context_;
+}
+
+EmbedderSurfaceVulkan::~EmbedderSurfaceVulkan() = default;
```

Comment:
See the note about object lifecycles later (in the test context). The GrContext needs to release all resources and abandon itself from the underlying handles.

Reply:
Fixed.